### PR TITLE
🎨 storage emits S3 file change events and dynamic-sidecar is notified when s3 content changes

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

- When operations are made on files in `storage`, a `FileNotificationEventType` is sent out.
- Currently only the `dynamic-sidecar` subscribes to these events.
  - if data mounting `enabled`: on every it trigger s directory reload (via rclone) to ensure the content on the disk is up to date with the one in S3.
  - if data mounting `disabled`: depending on the event type, the file will either be removed form the disk or downloaded form S3

## Related issue/s

- https://github.com/ITISFoundation/private-issues/issues/8

### CODE REVIEW

File: api/specs/web-server/_storage.py [hunk 1]
```diff
@@ -71,6 +72,14 @@ async def list_storage_paths(
     """Lists the files/directories in WorkingDirectory"""
 
 
+@router.post(
+    "/storage/locations/{location_id}/paths/{s3_directory}:refresh",
```
Comment (sanderegg) [inline api/specs/web-server/_storage.py]: what is s3_directory?
we have the concept of path in here. please define what you mean by this

File: packages/simcore-sdk/tests/unit/test_node_ports_common_r_clone_mount.py [hunk 2]
```diff
@@ -273,6 +297,18 @@ async def _get_file_checksums_from_s3(
     return checksums
 
 
+async def _create_file_in_s3_path(
+    s3_client: S3Client, bucket_name: S3BucketName, remote_path: StorageFileID, *, file_name: Path, file_size: ByteSize
+) -> tuple[Path, str]:
+    with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+        await _create_random_binary_file(Path(tmp_file.name), file_size)
+        await s3_client.upload_file(tmp_file.name, bucket_name, f"{remote_path}/{file_name}")
```
Comment (sanderegg) [inline packages/simcore-sdk/tests/unit/test_node_ports_common_r_clone_mount.py]: and this as well. please check what storage tests are doing. they for sure use some fixtures

**Review by sanderegg — COMMENTED** [review]
please check how the API for storage is done and keep using the same nomenclature there. no need to add confusing things.
Also what is the point of telling the frontend what type of mount we are using??

File: packages/simcore-sdk/tests/unit/test_node_ports_common_r_clone_mount.py [hunk 3]
```diff
@@ -190,8 +205,17 @@ def index() -> int:
 
 
 @pytest.fixture
-def remote_path(faker: Faker) -> StorageFileID:
-    return TypeAdapter(StorageFileID).validate_python(f"{faker.uuid4()}/{faker.uuid4()}/mounted-path")
+def ensure_remote_path(faker: Faker) -> Callable[[], StorageFileID]:
+    def _() -> StorageFileID:
+        remote = f"{faker.uuid4()}/{faker.uuid4()}/mounted-path"
```
[the remark this answers is not in the document]
Comment (GitHK) [reply packages/simcore-sdk/tests/unit/test_node_ports_common_r_clone_mount.py]: replied below

File: packages/models-library/src/models_library/api_schemas_webserver/storage.py [hunk 4]
```diff
@@ -28,6 +28,10 @@ class StoragePathComputeSizeParams(StorageLocationPathParams):
     path: Path
 
 
+class StoragePathRefreshParams(StorageLocationPathParams):
+    s3_directory: SimcoreS3FileID
```
[the remark this answers is not in the document]
Comment (GitHK) [reply packages/models-library/src/models_library/api_schemas_webserver/storage.py]: I've removed this now, it's not exposed in the fronted but still used internally

**Review by sanderegg — COMMENTED** [review]
I still would like to understand the use-case.

this is exposed to the frontend right? why is this only applicable to rclone mounts? what is the workflow? the frontend will call this?

File: packages/models-library/src/models_library/api_schemas_directorv2/dynamic_services_service.py [hunk 5]
```diff
@@ -56,33 +61,34 @@ class ServiceDetails(CommonServiceDetails):
...
-    state: ServiceState = Field(..., description="service current state", alias="service_state")
-    message: str | None = Field(
-        default=None,
-        description="additional information related to service state",
-        alias="service_message",
-    )
+    entry_point: Annotated[
+        str | None, Field(description="if empty the service entrypoint is on the root endpoint.", deprecated=True)
+    ] = None
+    state: Annotated[ServiceState, Field(description="service current state", alias="service_state")]
```
Comment (pcrespov) [inline packages/models-library/src/models_library/api_schemas_directorv2/dynamic_services_service.py]: Q: why using aliases and not just call it `service_state`? etc

File: packages/simcore-sdk/src/simcore_sdk/node_ports_common/r_clone_mount/_manager.py [hunk 6]
```diff
@@ -186,11 +192,30 @@ async def ensure_unmounted(self, local_mount_path: Path, index: NonNegativeInt)
...
+
+            for mount_id, remote in self._reverse_path_search.items():
+                if base_s3_path == remote:
+                    tracked_mount = self._tracked_mounts[mount_id]
+                    break
+
+            if tracked_mount is None:
+                raise NoMountFoundForRemotePathError(remote_path=remote_path)
+
+            dir_to_refresh = "/".join(remote_path_parts[3:])
```
Comment (pcrespov) [inline packages/simcore-sdk/src/simcore_sdk/node_ports_common/r_clone_mount/_manager.py]: MINOR: perhaps the split i.e. `3` you can put it in a constant and used in both paths i.e. here and above?

**Review by pcrespov — APPROVED** [review]
thx. left several comments

File: packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/dynamic_scheduler/services.py [hunk 7]
```diff
@@ -159,3 +159,18 @@ async def update_projects_networks(rabbitmq_rpc_client: RabbitMQRPCClient, *, pr
         timeout_s=_RPC_DEFAULT_TIMEOUT_S,
     )
     assert result is None  # nosec
+
+
+@log_decorator(_logger, level=logging.DEBUG)
+async def notify_path_change(
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/dynamic_scheduler/services.py]:
```suggestion
async def reload_path(
```

File: packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/dynamic_scheduler/services.py [hunk 8]
```diff
@@ -159,3 +159,18 @@ async def update_projects_networks(rabbitmq_rpc_client: RabbitMQRPCClient, *, pr
...
+    rabbitmq_rpc_client: RabbitMQRPCClient, *, node_id: NodeID, path: StorageFileID, recursive: bool = False
+) -> None:
+    result = await rabbitmq_rpc_client.request(
+        DYNAMIC_SCHEDULER_RPC_NAMESPACE,
+        _RPC_METHOD_NAME_ADAPTER.validate_python("notify_path_change"),
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/dynamic_scheduler/services.py]:
```suggestion
        _RPC_METHOD_NAME_ADAPTER.validate_python("reload_path"),
```

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/services/container_extensions.py [hunk 9]
```diff
@@ -97,3 +100,14 @@ async def detach_container_from_network(*, container_id: str, network_id: str) -
         # used by a container
         network = DockerNetwork(docker=docker, id_=network_id)
         await network.disconnect({"Container": container_id, "Force": True})
+
+
+async def notify_path_change(app: FastAPI, *, path: StorageFileID, recursive: bool) -> None:
```
Comment (sanderegg) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/services/container_extensions.py]: reload

File: services/web/server/src/simcore_service_webserver/dynamic_scheduler/api.py [hunk 10]
```diff
@@ -225,3 +225,8 @@ async def retrieve_inputs(
 
 async def update_projects_networks(app: web.Application, *, project_id: ProjectID) -> None:
     await services.update_projects_networks(get_rabbitmq_rpc_client(app), project_id=project_id)
+
+
+async def notify_path_change(app: web.Application, *, path: StorageFileID) -> None:
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/dynamic_scheduler/api.py]: reload

File: services/web/server/src/simcore_service_webserver/storage/_rest.py [hunk 11]
```diff
@@ -180,6 +182,27 @@ async def list_paths(request: web.Request) -> web.Response:
...
+    # If no service is running, no errors are raised.
+    _ = AuthenticatedRequestContext.model_validate(request)
+    path_params = parse_request_path_parameters_as(StoragePathNotifyChangeParams, request)
+
+    await dynamic_scheduler.notify_path_change(
+        request.app, path=TypeAdapter(StorageFileID).validate_python(f"{path_params.path}")
+    )
+
+    return web.json_response(status=status.HTTP_204_NO_CONTENT)
+
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/storage/_rest.py]: reload

**Review by sanderegg — COMMENTED** [review]
I still don't get why this reloading is only needed for rclone mounts and not the other modes?

Also please update the PR description. thanks

**Review by sanderegg — CHANGES_REQUESTED** [review]
This is getting better but:
- why is this only implemented for the rclone backend, that is not even yet the default?
- please check the code in storage, do not duplicate re-invent code for parsing file ids

File: packages/models-library/src/models_library/api_schemas_directorv2/dynamic_services_service.py [hunk 12]
```diff
@@ -56,37 +61,34 @@ class ServiceDetails(CommonServiceDetails):
...
+        ServiceBootType,
+        Field(
+            description=(
+                "Describes how the dynamic services was started (legacy=V0, modern=V2)."
+                "Since legacy services do not have this label it defaults to V0."
+            )
         ),
-    )
+    ] = ServiceBootType.V0
+    data_mounting_enabled: Annotated[bool, Field(description="True if data mounting is enabled")] = False
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/api_schemas_directorv2/dynamic_services_service.py]: why? I thought we discussed that it makes no sense...

File: packages/models-library/src/models_library/basic_regex.py [hunk 13]
```diff
@@ -42,7 +42,13 @@
 MIME_TYPE_RE = r"([\w\*]*)\/(([\w\-\*]+\.)+)?([\w\-\*]+)(\+([\w\-\.]+))?(; ([\w+-\.=]+))?"
 
 # Storage basic file ID
-SIMCORE_S3_FILE_ID_RE = rf"^(exports\/\d+\/{UUID_RE_BASE}\.zip)|((api|({UUID_RE_BASE}))\/({UUID_RE_BASE})\/(.+)$)"
+_ALLOWED_PREFIX_EXPORTS: Final[str] = "exports"
+_ALLOWED_PREFIX_API: Final[str] = "api"
+SIMCORE_S3_FILE_ID_ALLOWED_PREFIXES: Final[set[str]] = {_ALLOWED_PREFIX_API, _ALLOWED_PREFIX_EXPORTS}
+SIMCORE_S3_FILE_ID_RE = (
```
Comment (sanderegg) [inline packages/models-library/src/models_library/basic_regex.py]: not sure I find that any easier to read ;)

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_scheduler_utils.py [hunk 14]
```diff
@@ -71,6 +71,7 @@ def create_model_from_scheduler_data(
     return RunningDynamicServiceDetails.model_validate(
         {
             "boot_type": ServiceBootType.V2,
+            "data_mounting_enabled": scheduler_data.requires_data_mounting,
```
Comment (sanderegg) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_scheduler_utils.py]: so here it makes sense since it's kind of internal... but to the frontend this has no meaning.
I do not understand why this should be exclusive to rclone mounts... makes no sense.

**Review by sanderegg — COMMENTED** [review]
Still do not see why only rclone mounts are capable of refreshing and not others.
and I don't see why the frontend should know what is going on in the backend anyway.
probably a feature like `allow live uploads` would be a better one.
and what about the delete button? are you going to remove it now?

File: packages/models-library/src/models_library/rabbitmq_messages.py [hunk 15]
```diff
@@ -326,3 +326,26 @@ class ComputationalPipelineStatusMessage(RabbitMessageBase, ProjectMessageBase):
...
+    project_id: ProjectID | None = None
+    node_id: NodeID | None = None
+    file_id: SimcoreS3FileID
+    created_at: datetime.datetime = Field(
+        default_factory=lambda: arrow.utcnow().datetime,
+        description="message creation datetime",
+    )
+
+    def routing_key(self) -> str | None:
+        return f"{self.project_id}.{self.node_id}"
```
[the remark this answers is not in the document]
Comment (GitHK) [reply packages/models-library/src/models_library/rabbitmq_messages.py]: notification sending is skipped for files with `api` or `exports` prefix, so this portion of the code is not reached

**Review by pcrespov — COMMENTED** [review]
can you pleaes reassign when you address @sanderegg points and get the CI green? At that point i think my review will be more effective. thx :-)

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+                exclusive_queue=True,
+                topics=[topic],
+            )
+            app.state.file_notification_queue = subscribed_queue
+
+    async def _stop() -> None:
+        queue_name: str = app.state.file_notification_queue
+        with log_context(_logger, logging.INFO, msg=f"unsubscribing from file notifications with queue={queue_name}"):
+            rabbit_client: RabbitMQClient = get_rabbitmq_client(app)
+            await rabbit_client.unsubscribe(queue_name)
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py]: this is a valid point

File: services/storage/src/simcore_service_storage/modules/rabbitmq.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+from servicelib.logging_utils import log_catch, log_context
+from servicelib.rabbitmq import RabbitMQClient
+
+from .._meta import APP_NAME
+from ..core.settings import get_application_settings
+
+_logger = logging.getLogger(__name__)
+
+
+def setup_rabbitmq(app: FastAPI) -> None:
```
Comment (pcrespov) [inline services/storage/src/simcore_service_storage/modules/rabbitmq.py:23]: at this point it is preferable to use `packages/service-library/src/servicelib/fastapi/rabbitmq_lifespan.py` for these instead of copy&pasting this structure around. WQhat do you think?

File: services/storage/tests/unit/test_modules_rabbitmq.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+async def test_post_file_notification_skipped_prefixes(mock_app: AsyncMock, file_id: str, faker: Faker):
+    await post_file_notification(
+        mock_app,
+        event_type=FileNotificationEventType.FILE_DELETED,
+        user_id=7,
+        file_id=file_id.format(uuid=faker.uuid4()),
+    )
+
+    mock_client = mock_app.state.rabbitmq_client
+    mock_client.publish.assert_not_called()
```
Comment (pcrespov) [inline services/storage/tests/unit/test_modules_rabbitmq.py]: MINOR: IMO  is more readable since the assertion is one as the other i.e. using `assert` 
```python
assert not mock_client.publish.called
```

File: services/storage/tests/unit/test_modules_rabbitmq.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+async def test_post_file_notification_standard_file_id(mock_app: AsyncMock, project_id: ProjectID, node_id: NodeID):
+    file_id = f"{project_id}/{node_id}/data.csv"
+
+    await post_file_notification(
+        mock_app,
+        event_type=FileNotificationEventType.FILE_UPLOADED,
+        user_id=42,
+        file_id=file_id,
+    )
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/storage/tests/unit/test_modules_rabbitmq.py:44]: rhis is a valid point as well

File: services/storage/tests/unit/test_modules_rabbitmq.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+    mock_client = mock_app.state.rabbitmq_client
+    mock_client.publish.assert_called_once()
```
Comment (pcrespov) [inline services/storage/tests/unit/test_modules_rabbitmq.py]: MINOR: as before, i think it is more readable to use `assert` to verify instead of custom functions

```python
assert mock_client.publish.call_count == 1
```

File: services/storage/src/simcore_service_storage/modules/rabbitmq.py [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+
+        if parts[0] in SIMCORE_S3_FILE_ID_ALLOWED_PREFIXES:
+            _logger.info("Skip notification for file_id=%s starting with prefix %s", file_id, parts[0])
+            return
+
+        project_id = ProjectID(parts[0]) if len(parts) > 0 else None
+        node_id = NodeID(parts[1]) if len(parts) > 1 else None
+
+        if project_id is None or node_id is None:
+            _logger.info(
```
Comment (pcrespov) [inline services/storage/src/simcore_service_storage/modules/rabbitmq.py]: what scenario allows this? if it is nto normal, should it log an error? or at least a warning.

File: services/storage/src/simcore_service_storage/modules/rabbitmq.py [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+        parts = f"{file_id}".split("/")
+        project_id = _try_parse_uuid(parts[0]) if len(parts) > 0 else None
+        node_id = _try_parse_uuid(parts[1]) if len(parts) > 1 else None
+
+        message = FileNotificationMessage(
+            event_type=event_type,
+            user_id=user_id,
+            project_id=project_id,
+            node_id=node_id,
+            file_id=file_id,
```
[the remark this answers is not in the document]
Comment (pcrespov) [reply services/storage/src/simcore_service_storage/modules/rabbitmq.py:98]: make at least a defensive `assert` above  to document that this is really going to be SimcoreS3FileID

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+
+    app.add_event_handler("startup", _startup)
```
Comment (pcrespov) [inline services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py:167]: same story here with `lifespans` ... please let's start converting to `lifepan` instead of start/stop pattern . see https://fastapi.tiangolo.com/advanced/events/

**Review by pcrespov — COMMENTED** [review]
left some comments. In order to speed up merging this PR, we can follow up offline if you need some input/help

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py (hunk 23 above)
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py:167]: Converting to lifespans is not going to be in the middle of a this pull request. I'd like for this to be a maintenance case, since it can get pretty involved.

Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py:167]: We can piggy back this existing case and expand on it https://github.com/ITISFoundation/osparc-simcore/issues/4678 or open a new one dedicated, like it was done for the pedantic or other systems

File: services/storage/src/simcore_service_storage/modules/rabbitmq.py (hunk 17 above)
Comment (GitHK) [reply services/storage/src/simcore_service_storage/modules/rabbitmq.py:23]: Not with the setup infrastructure, with the lifespan it's easier because we can inject an expected context where the settings are present, we already have examples of this. So I would wait for the lifespans PR to change all these

File: services/storage/src/simcore_service_storage/modules/rabbitmq.py (hunk 21 above)
Comment (GitHK) [reply services/storage/src/simcore_service_storage/modules/rabbitmq.py]: Good one, warning is better since it's a safety net for when we add something in the future.

File: services/storage/tests/unit/test_modules_rabbitmq.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (GitHK) [reply services/storage/tests/unit/test_modules_rabbitmq.py:44]: I don't agree with this statement, this is actually the better edge case, it works even in the root folder, otherwise the test would have failed.
Also the node_id and project_id are validated by the function and an error would have been raised.

[the remark this answers is not in the document]
Comment (GitHK) [reply services/storage/tests/unit/test_modules_rabbitmq.py:44]: I've also extended the tests (while fixing an issue) to also include a few more cases that would 100% avoid this

File: packages/simcore-sdk/src/simcore_sdk/node_ports_common/r_clone_mount/_manager.py [hunk 24]
```diff
@@ -197,11 +206,31 @@ async def ensure_unmounted(self, local_mount_path: Path, index: NonNegativeInt)
...
+            self._reverse_path_search.pop(mount_id)
 
             await tracked_mount.wait_for_all_transfers_to_complete()
 
             await tracked_mount.stop_mount()
 
+    async def refresh_path(self, remote_path: StorageFileID, *, recursive: bool = False) -> None:
+        with log_context(_logger, logging.INFO, f"refreshing mount for {remote_path=}", log_duration=True):
+            remote_path_parts = remote_path.split("/")
+            assert len(remote_path_parts) >= _MIN_PATH_PARTS, "Expected {project_id}/{node_id}/DIRECTORY_PATH"
```
[the remark this answers is not in the document]
Comment (GitHK) [reply packages/simcore-sdk/src/simcore_sdk/node_ports_common/r_clone_mount/_manager.py:218]: this is safe

File: services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+    except NoMountFoundForRemotePathError:
+        mounted_volumes: MountedVolumes = app.state.mounted_volumes
+        match event_type:
+            case FileNotificationEventType.FILE_UPLOADED:
+                await _try_pull_to_disk_volumes(app, mounted_volumes, path)
+            case FileNotificationEventType.FILE_DELETED:
+                await _try_remove_from_disk_volumes(mounted_volumes, path)
+            case _:
+                _logger.warning("Received unsupported event type '%s' for path '%s'", event_type, path)
+
```
[the remark this answers is not in the document]
Comment (GitHK) [reply services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/file_notification_subscriber.py:135]: not interested in those it's OK if they bubble up

File: services/storage/src/simcore_service_storage/modules/rabbitmq.py (hunk 17 above)
Comment (pcrespov) [reply services/storage/src/simcore_service_storage/modules/rabbitmq.py:23]: ok, what about creating an issue (under maintenance) to push forwards lifespans?

Comment (pcrespov) [reply services/storage/src/simcore_service_storage/modules/rabbitmq.py:23]: ok, i saw https://github.com/ITISFoundation/osparc-simcore/issues/4678

**Review by pcrespov — APPROVED** [review]
thx

**Review by matusdrobuliak66 — APPROVED** [review]
Approve to unblock.

**Review by giancarloromeo — APPROVED** [review]
